### PR TITLE
Hyperbolic tangent instability

### PR DESCRIPTION
**Pull request by spetravic**
The current implementation of the action sampling using _RobustTanhTransform_ causes some numerical instability. This leads to the log probability getting extreme values, which then completely destabilizes the training and the creature fails to learn.

image
image

_TanhTransform_ with cache_size=1 caches the original action values sent to the transform and returns them instead of doing the inverse transformation, which seems to prevent the instability issue. The results are in line with Cog training runs, and the creature is able to learn the tasks.

image
image

### CODE REVIEW

**Review by AliGhadirzadeh — APPROVED** [review]
This looks good to me. It should speed up the policy inference as it does not compute tanh and atanh for every single state. Also, gradients are not passed through (a)tanh functions during bp step. Very nice :)

File: emote/nn/gaussian_policy.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def forward(
         if self.training:
             dist = dists.TransformedDistribution(
                 dists.Independent(dists.Normal(mean, std), 1),
-                RobustTanhTransform(),
+                transforms.TanhTransform(cache_size=1),
```
Comment (tgolsson) [inline emote/nn/gaussian_policy.py:70]: You can do this with the RobustTanhTransform too, it derives from this class.

Comment (spetravic) [reply emote/nn/gaussian_policy.py:70]: I figured since the values are cached and the transformation is not actually inverted, it doesn't make much of a difference (if I am not misunderstanding how the cache works).

**Review by singhblom — APPROVED** [review]

File: emote/nn/gaussian_policy.py (hunk 1 above)
Comment (spetravic) [reply emote/nn/gaussian_policy.py:70]: I don't know whether to laugh or cry about that, to be honest 🥲

Comment (AliGhadirzadeh) [reply emote/nn/gaussian_policy.py:70]: I do not think we still need RobustTanhTransform. We do not need to clamp the output of tanh anymore. Even for very large inputs to tanh, the transformation still caches the input and returns the right value for loglikelihood calculations.

Comment (tgolsson) [reply emote/nn/gaussian_policy.py:70]: What happens if we're running forward and backward passes in parallel though? That's the case for clients which doesn't do local inference during training. In that case the cache would not be valid.

Comment (singhblom) [reply emote/nn/gaussian_policy.py:70]: Wait, how would the cache not be valid? It's always used on the line after it's set, right?

```python
# Transform created
dist = dists.TransformedDistribution(
    dists.Independent(dists.Normal(mean, std), 1),
    transforms.TanhTransform(cache_size=1),
)
# Transform called, cache set
sample = dist.rsample()
# Inverse called in log_prob, cache used
log_prob = dist.log_prob(sample).view(bsz, 1)
```

Comment (singhblom) [reply emote/nn/gaussian_policy.py:70]: But I do agree with Ali, if we don't use the RobustTanh we should remove it.

Comment (tgolsson) [reply emote/nn/gaussian_policy.py:70]: Ah, it's local. Then yeah nvm me. :D Keep thinking like tensorflow with lots of global state.
